K8s on desktop demo doc (#201)
LiqoTech/liqo: docs/pages/User/GettingStarted/demo_k8s_on_desktop.md

## Installation of the required software
To install all the required software we need to follow this steps:

1. [Install k3s](#install-k3s) in both clusters;
2. [Install liqo.io](#install-liqo-io) in both clusters; 
3. [Install KubernetesOnDesktop](#install-kubernetesondesktop) just in one of the clusters (the so called *local cluster*).

### Install k3s
Assuming you already have two linux (we tested it with Ubuntu Desktop 20.04 LTS) machine or Virtual Machine up and running in the same subnet, we can install [k3s](https://k3s.io/) by using the official script as you can see in [K3s Quick-Start Guide](https://rancher.com/docs/k3s/latest/en/quick-start/). So, you only need to run the following:

```bash
curl -sfL https://get.k3s.io | sh -
```

When the script ends, to make [liqo.io](https://liqo.io) work properly, you need to modify the `/etc/systemd/system/k3s.service` file by adding the `--kube-apiserver-arg anonymous-auth=true` service execution parameter. You can do this by executing the following:

```bash
sudo sed -i "s#server#server --kube-apiserver-arg anonymous-auth=true#" /etc/systemd/system/k3s.service
```

> After this operation, your `k3s.service` file should look like this:

Reviewer: An [expand](https://learn.netlify.app/en/shortcodes/expand/) could be useful here.

Author: Fixed

[Service]
Type=notify
EnvironmentFile=/etc/systemd/system/k3s.service.env
KillMode=process
Delegate=yes
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server --kube-apiserver-arg anonymous-auth=true \

```

Now you need to apply the changes by executing the following:

```bash
systemctl daemon-reload
systemctl restart k3s.service
```

Finally, to use `kubectl` command to interact with the cluster without `sudo`, you need to give the `k3s.yaml` config file the required privileges and to copy it in `~/.kube/config` by doing the following:

```bash
sudo chmod 666 /etc/rancher/k3s/k3s.yaml
cp /etc/rancher/k3s/k3s.yaml ~/.kube/config
```

Before proceeding with the [liqo.io](https://liqo.io) installation, wait for all the pod to be in `Running` status. You can check it by executing `kubectl get pod --all-namespaces`.

> ### Install liqo.io

Reviewer: just Liqo, without io

Author: Fixed

#### About KubernetesOnDesktop

[KubernetesOnDesktop](https://github.com/netgroup-polito/KubernetesOnDesktop) is a University project with the aim of developing a cloud infrastructure to run user application in a remote cluster node. 
It uses a client/server VNC+PulseAudio+SSH infrastructure that schedules the application `pod` in a k8s remote node and redirects the GUI (through VNC) and the sound (through PulseAudio+SSH) in a second `pod` scheduled in the node in which the `cloudify` application is running. For further information see [KubernetesOnDesktop](https://github.com/netgroup-polito/KubernetesOnDesktop) GitHub page.

So far, the supported applications are firefox, libreoffice and blender (with the ability to use NVIDIA CUDA driver if the node which the `pod` will be executed in has a NVIDIA graphic card). Anyway, thanks to a huge use of templates, it is possible to scale up to many more applications.

This project uses several kubernetes components, such as `deployments`, `jobs`, `services` (with `ClusterIP` and `NodePort` `type` values) and `secrets`. All those components, as you can see afterwards, are supported by [liqo.io](https://liqo.io).

> When executed, through `cloudify` command, the application will create:

Reviewer: Alsd candidate for "Expand"

Author: Fixed

```bash
curl -sfL https://get.k3s.io | sh -
```

A tiny customization of the above default install is required for [liqo](https://liqo.io) to work.
When the script ends, you need to modify the `/etc/systemd/system/k3s.service` file by adding the `--kube-apiserver-arg anonymous-auth=true` service execution parameter as in the following command:

```bash
sudo sed -i "s#server#server --kube-apiserver-arg anonymous-auth=true#" /etc/systemd/system/k3s.service
```

{{%expand "After this operation, your `k3s.service` file should look like the following:" %}}

```
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
Wants=network-online.target

[Install]
WantedBy=multi-user.target

[Service]
Type=notify
EnvironmentFile=/etc/systemd/system/k3s.service.env
KillMode=process
Delegate=yes
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \

>     server --kube-apiserver-arg anonymous-auth=true \

Reviewer 2: if you want to run it on nVidia docker, you have to add also a `--docker` flag to enable k3s to run on docker

Author: Fixed

So far, KoD suppors firefox, libreoffice and blender, the latter with the capability to exploit any available NVIDIA GPUs (through the NVIDIA CUDA driver) if the remote node has this hardware.
In any case, thanks to the massive use of templates, many more applications can be easily supported.

When executing the `cloudify` command, the application will create:
  * a `secret` containing an ssh key that allows the two application components to communicate securely;
  * a `deployment` containing the application (e.g. blender) and the VNC server, whose `pod` will be scheduled on the remote cluster;
  * a `ClusterIP` `service` that allows the remote `pod` to be reachable from other `pod`s in the cluster by using [K8s DNS for services](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/);
  * a `pod` executing the VNC viewer, started in the local machine (i.e., on the same node where you run `cloudify`).
  
## Installation of the required software
To install all the required software we need to follow this steps:

1. [Install k3s](#install-k3s) in both clusters;

> 2. [Install liqo](#install-liqo) in both clusters; 

Reviewer 3: @riccardoroccaro It seems to me that we always wrote liqo with the capital letter (i.e., Liqo). Please inspect this file and change all occurrencies of "liqo" accordingly (only when liqo is a name, not part of something else, e.g., liqo-<...> such as the name of the pod). Thanks!

Author: Fixed

When the script ends, you need to modify the `/etc/systemd/system/k3s.service` file by adding the `--kube-apiserver-arg anonymous-auth=true` service execution parameter as in the following command:

```bash
sudo sed -i "s#server#server --kube-apiserver-arg anonymous-auth=true#" /etc/systemd/system/k3s.service
```

{{%expand "After this operation, your `k3s.service` file should look like the following:" %}}

```
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
Wants=network-online.target

[Install]
WantedBy=multi-user.target

[Service]
Type=notify
EnvironmentFile=/etc/systemd/system/k3s.service.env
KillMode=process
Delegate=yes
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server --kube-apiserver-arg anonymous-auth=true \

```
{{% /expand%}}

> {{%expand "**Note**: If you want to exploit any available NVIDIA GPUs in the *foreign cluster* you also have to do in that cluster as follows:" %}}

Reviewer 3: ```suggestion
{{%expand "**Note**: If you want to exploit any available NVIDIA GPUs in the *foreign cluster*, you have to follow the additional steps below (in the foreign cluster):" %}}
```

Author: Fixed

1. [Install in the required NVIDIA CUDA driver](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#how-do-i-install-the-nvidia-driver);

> 2. [Install the Docker engine](https://docs.docker.com/engine/install/);

Reviewer 3: Is this needed? I suppose that Docker is already installed when you setup K3s. Am I wrong?

Author: Actually k3s uses containerd instead of docker, so we need to install it on our own.

3. [Install the `nvidia-container-runtime`](https://github.com/nvidia/nvidia-container-runtime#installation);

> 4. Add the `--docker` service execution parameter in the `/etc/systemd/system/k3s.service` file by executing the following command:

Reviewer 3: Add the `--docker` service execution parameter ==> Add the `--docker` service execution parameter TO DO WHAT?
Tell people not only what to do, but also why is this needed (briefly, very briefly)

Author: Fixed
